Add support for DOIO KB30-01

crkochan opened the pull request:
Adds support for [the DOIO KB30-01 "Megalodon" 30% keyboard from Drop](https://drop.com/buy/megalodon-doio-hot-swappable-rgb-30-keyboard).

crkochan committed:
Lower OLED timeout to 1 minute

crkochan committed:
Fix direction of the large encoder knob

crkochan committed:
Support for DOIO KB30-01

crkochan committed:
Reduce firmware size

lesshonor commented on keyboards/doio/kb30/config.h:
Please revert changes to this file. It will cause conflicts when this keyboard is updated upstream in QMK.
If you want a shorter timeout on the Vial keymap, put it on the Vial keymap. If you think it's a better idea for the entire keyboard, submit this change upstream.

crkochan replied:
Reverted locally.

TIL about `#undef`.

lesshonor commented on keyboards/doio/kb30/info.json:
Please revert changes to this file. The Vial keymap submission is not an appropriate place to modify pin definitions for boards that were submitted to and still exist in upstream QMK.

To summarize [comments I left on a previous PR](https://github.com/vial-kb/vial-qmk/pull/473#discussion_r1209481555):
- if the encoders are wrong for everyone, you should fix them for everyone (i.e.: in QMK)
- if your particular encoder is backwards because you sourced it yourself vs using the included/intended hardware, add `#define ENCODER_DIRECTION_FLIP` to your keymap's `config.h`
- if the manufacturer couldn't be bothered to source consistent hardware so different runs of the keyboard use incompatible encoders...I'd probably just shrug my shoulders and let people fend for themselves. But you could consider adding additional layouts that allow end-users to flip the encoder directions in the GUI.

crkochan replied:
The QMK documentation on defining encoders finally clicked, so I was able to suss out how to implement the reverse just for vial.

`ENCODER_DIRECTION_FLIP` wasn't a good fix because only one of the three knobs on the device is backwards.

lesshonor commented on keyboards/doio/kb30/keymaps/vial/config.h:
```diff
@@ -0,0 +1,15 @@
+#define VIAL_UNLOCK_COMBO_COLS { 0, 3 }
+
+#define ENCODER_MAP_ENABLE
+
```
```suggestion
```

lesshonor commented on keyboards/doio/kb30/keymaps/vial/rules.mk:
```diff
@@ -0,0 +1,7 @@
+
+LTO_ENABLE = yes
+
+ENCODER_ENABLE = yes
```
```suggestion
ENCODER_ENABLE = yes
ENCODER_MAP_ENABLE = yes
```

crkochan commented:
I am working on incorporating the feedback, but I won't have access to the hardware again until Monday, and I would prefer to test the changes before pushing the changes to the PR.

crkochan committed:
Revert changes to base files

crkochan committed:
Override base OLED timeout

crkochan committed:
Move encoder map enable to rules.mk

crkochan committed:
Merge pull request #1 from crkochan/doio-kb30-corrections

Doio kb30 corrections

crkochan committed:
Pin swap encoder 3 in vial.

Large encoder otherwise acts reverse of how it is defined in the keymap.c file and the Vial configurator.
